### PR TITLE
Added documentation for abstract syntax

### PR DESCRIPTION
**Pull request by EnzeXing**
Added the abstract syntax for the data structure for abstract domain used in `Object.scala` in PR #16970

### CODE REVIEW

**Review by olhotak — APPROVED** [review]

File: compiler/src/dotty/tools/dotc/transform/init/Objects.scala [hunk 1]
```diff
@@ -65,6 +65,34 @@ import scala.annotation.constructorOnly
...
+ *  env = (valsMap, Option[env])                                        // stores local variables in the residing method, and possibly outer environments
+ *
+ *  addr ::= localVarAddr(regions, valsym, owner)
+ *  | fieldVarAddr(regions, valsym, owner)                              // independent of OfClass/ObjectRef
+ *  | arrayAddr(regions, owner)                                         // independent of array element type
+ *
+ *  regions ::= List(sourcePosition)
+ *
+ *  Note: -> refers to extensible maps whose entries cannot be modified; -M> refers to extensible whose entries can be modified
+ */
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/transform/init/Objects.scala:95]: We need to indent this properly. Maybe send a PR to `lampepfl/dotty` later after we merge the main PR.

File: compiler/src/dotty/tools/dotc/transform/init/Objects.scala [hunk 2]
```diff
@@ -65,6 +65,34 @@ import scala.annotation.constructorOnly
 object Objects:
 
   // ----------------------------- abstract domain -----------------------------
+/** Syntax for the data structure abstraction used in abstract domain:
+
+ *  ve ::= ObjectRef(class)
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/transform/init/Objects.scala:70]: We miss a star for the blank line.

File: compiler/src/dotty/tools/dotc/transform/init/Objects.scala [hunk 3]
```diff
@@ -65,6 +65,34 @@ import scala.annotation.constructorOnly
...
+ *  | OfClass(class, vs[outer], ctor, args, env)
+ *  | OfArray(object[owner], regions)
+ *  | Fun(..., env)
+ *  | Cold // abstract values in domain
+ *  vs ::= Set(ve) // set of abstract values
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/transform/init/Objects.scala:75]: Minor: align comments vertically.

**Comment by liufengyun** [comment]
BTW, it seems the PR is made from `dotty-staging/dotty-1`. Is that accidentally created? If that's the case, it's better to remove that repo, as it will be confusing.

I'd suggest close this PR and piggyback the document to the PR you are going to make about the domain changes.

**Comment by EnzeXing** [comment]
@liufengyun I forked the repo from `dotty-staging/dotty`. Shall I fork the official repo instead?

**Comment by liufengyun** [comment]
> @liufengyun I forked the repo from `dotty-staging/dotty`. Shall I fork the official repo instead?

It's not supposed to create any new fork of dotty under `dotty-staging`. You can create a branch in `dotty-staging/dotty` directly. 

The repo `dotty-staging/dotty` is not intended for PRs. I suggest we close this for now and make a new PR to `lampepfl/dotty` once https://github.com/lampepfl/dotty/pull/16970 is merged.

**Pull request closed without merging**